### PR TITLE
Dictionary view

### LINKED ISSUE #1357: Map: display

### PR DESCRIPTION
**Pull request by steffenagger**
## Warning:

Dictionary support is currently not published. These updated currently requires manually linking https://github.com/realm/realm-js/pull/3631 for testing.

This PR includes a generic "Json Viewer" modal, for displaying JSON in a popup.

This closes #1357

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -24,3 +24,6 @@ spectron.log
 
 # Sentry cache folder
 .electron-symbols/
+
+# Products
+*.tgz
```
Comment (kraenhansen) [inline .gitignore:29]: What might produce these?

File: src/ui/RealmBrowser/Content/Table/HeaderCell.tsx [hunk 2]
```diff
@@ -41,17 +43,42 @@ const getPropertyType = (property: Realm.ObjectSchemaProperty) => {
...
+    case 'dictionary':
+      return '{}';
+
+    case 'set':
+      return '<>';
+
+    default:
+      return '';
+  }
+};
```
Comment (kraenhansen) [inline src/ui/RealmBrowser/Content/Table/HeaderCell.tsx:69]: Just FYI. I hope we can decide to change these before the official release.

File: src/ui/RealmBrowser/Content/Table/HeaderCell.tsx [hunk 3]
```diff
@@ -41,17 +43,42 @@ const getPropertyType = (property: Realm.ObjectSchemaProperty) => {
   }
 };
 
+const optionalMarkOptOut = ['list', 'dictionary', 'set'];
```
Comment (kraenhansen) [inline src/ui/RealmBrowser/Content/Table/HeaderCell.tsx:46]:
```suggestion
const NON_OPTIONAL_TYPES = ['list', 'dictionary', 'set'];
```
Following the pattern used below.

File: src/ui/RealmBrowser/Content/Table/types/DictionaryCell.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+import React from 'react';
+import { Badge } from 'reactstrap';
+import Realm from 'realm';
+import { asSafeJsonString } from '../../../../../utils/json';
+
+import * as primitives from '../../../primitives';
+
+// TODO: Get declaration from Realm
```
Comment (kraenhansen) [inline src/ui/RealmBrowser/Content/Table/types/DictionaryCell.tsx:26]: Let's remember this once an official release with dictionaries is out 👍

File: src/utils/json.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Regex to clean up added $refId & $ref properties, added by Realm.JsonSerializationReplacer
+// TODO: Investigate better solution.
```
Comment (kraenhansen) [inline src/utils/json.ts:2]: Will this be fixed before merging the PR? Perhaps you need to enable passing an option to the `JsonSerializationReplacer`?

File: src/utils/json.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        indentation,
+      );
+    } catch (err) {
+      json = err.message ? err.message : err;
+    }
+  }
+
+  if (options) {
+    if (options.cleanupRefs) {
+      json = json.replace($REF_MATCHER, '');
```
Comment (kraenhansen) [inline src/utils/json.ts:38]: This will look a bit broken in the output won't it?

**Review by kraenhansen — APPROVED** [review]
Once RJS is released with dictionary support, this can be updated and merged.

File: src/ui/RealmBrowser/Content/Table/types/DictionaryCell.tsx (hunk 4 above)
Comment (steffenagger) [reply src/ui/RealmBrowser/Content/Table/types/DictionaryCell.tsx:26]: Yeah, they're not in realm-js just yet.

File: src/utils/json.ts (hunk 5 above)
Comment (steffenagger) [reply src/utils/json.ts:2]: That was also my initial thought, but this is in place if it's decided not to extend `JsonSerializationReplacer`.

File: src/utils/json.ts (hunk 6 above)
Comment (steffenagger) [reply src/utils/json.ts:38]: For what I tested, it looked as intended. There might be edge-cases, where the solution mentioned in the previous comment would avoid any/all edge-cases.

File: src/ui/RealmBrowser/Content/Table/HeaderCell.tsx (hunk 2 above)
Comment (steffenagger) [reply src/ui/RealmBrowser/Content/Table/HeaderCell.tsx:69]: Are we talking about `string[]` => `List<string>`? If so, that needs to be corrected several places.

Comment (kraenhansen) [reply src/ui/RealmBrowser/Content/Table/HeaderCell.tsx:69]: Yep - at least that's my hope 🤞

**Comment by steffenagger** [comment]
Closing this one. Will create a combined Set, Dictionary & Mixed PR.

**Pull request closed without merging**